### ISSUE TITLE
Drop `testtools` support?

### ISSUE THREAD

**Issue by vytas7** [falconry/falcon#2156]

**Comment by vytas7** [comment]
Another option is to drop `testtools`, but leave an escape hatch via an envvar along the lines of [Custom HTTP Methods](https://falcon.readthedocs.io/en/stable/api/routing.html#custom-http-methods) if anyone is still using that.

**Comment by CaselIT** [comment]
I have no opinion on this, either way works for me

**Comment by kgriffs** [comment]
I don't think we need to keep supporting it. It's a minor convenience left over from early OpenStack days for a very small community relative to the broader python ecosystem.

**Comment by vytas7** [comment]
We won't be able to fully drop it until 5.0, but we should emit a deprecation warning whenever the `TestCase` is rebased on top of `testtools.TestCase`, and add an escape hatch via an envvar to be able to import a custom testcase base.

**Comment by EricGoulart** [comment]
I’m sorry for the mistake with my previous pull request. I realize now that this wasn’t the correct process, and I apologize for any inconvenience it may have caused. I’m currently in the process of fixing this and will submit the pull request the correct way.

**Comment by vytas7** [comment]
No worries @EricGoulart, and thanks for your interest in contributing to this issue!
These mistakes do no real harm except a notification or two in our mailboxes, which is not the end of the world :sweat_smile:

**Comment by vytas7** [comment]
Note that using `testtools` actually creates a real bug, uncovered by recent versions of Mypy:
```
falcon/testing/test_case.py:34: error: Definition of "patch" in base class "TestCase" is incompatible with definition in base class "TestClient"  [misc]
```

Falcon's [`TestClient.patch`](https://falcon.readthedocs.io/en/stable/api/testing.html#falcon.testing.TestClient.patch) is shadowed by `testtools.TestClient.patch` via diamond inheritance... I guess we need to keep the `testtools` flavour for now, since it was introduced prior to the `patch()` alias (of `simulate_patch`).

**Comment by CaselIT** [comment]
I think it's fine to remove for v5

**Comment by ApparentlyTejas** [comment]
I'd like to take this. Rather than a full deprecation (which touches the public dependency surface and seems worth a separate v5 discussion), I'm going with the minimal fix @vytas7 outlined as a fallback: reword the `TestCase` docstring so it doesn't read as recommending `testtools`, and fix the underlying mypy issue.

Root cause: `falcon/testing/test_case.py` does `import testtools as unittest` / `except ImportError: import unittest`. When `testtools` is installed, mypy resolves both branches and raises `Name "unittest" already defined (by an import) [no-redef]`. I've reproduced this locally with `testtools` installed.

Fix: gate the import behind `TYPE_CHECKING` so type checkers always see `unittest` (matching runtime behavior when `testtools` isn't installed), while runtime still prefers `testtools` when present. This also lets me drop the now-unused `testtools` entry from the mypy `ignore_missing_imports` overrides and the now-unnecessary `# type: ignore[misc]` on the `TestCase` class.

Opening a PR shortly.
